feat(scanner): implement vuln update pipeline

RTann opened the pull request:
## Description

Initial implementation of Scanner-side vulnerability updating. These changes consist of:

* Implementation of the `GetMetadata` Matcher RPC
  * Querying for the last update time is required for the updater to run successfully, so I figured I might as well throw this in
* `MatcherMetadataStore`
  * This is the datastore used to get Matcher metadata. Currently, it consists of just a single table: `last_vuln_update`
  * This comes with an initialization and migration of related PostgreSQL tables
* Utilizing `libvuln.OfflineImport`
  * Currently, I use a test endpoint to fetch all the vulnerabilities. Once Central and Sensor are both implemented to support this, we can switch it over to something configurable
* DB integration tests
  * Added a new make target, `db-integration-test` to run tests which require a PostgreSQL connection

Note: I realized our original plan of just relying on the scanner version to determine which vuln data to use only works for releases. This is not very usable for testing. I'm thinking we need to take time to determine how to handle testing (probably via some `master` branch version of data)

Resolves: https://issues.redhat.com/browse/ROX-20355

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

### Here I tell how I validated my change

Aside from the unit tests/integration test, I have this running in an OpenShift 4 cluster.

Follow the steps from https://github.com/stackrox/stackrox/pull/8869. Note: `secret_values.yaml` is the following:

```
imagePullSecrets:
  username: <Your quay creds>
  password: <Your quay creds>
```

For test workloads I used, see https://github.com/stackrox/stackrox/pull/9101

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

RTann commented on scanner/matcher/matcher.go:
```diff
@@ -40,43 +44,69 @@ func NewMatcher(ctx context.Context, cfg config.MatcherConfig) (Matcher, error)
+		Pool:          pool,
+		MetadataStore: metadataStore,
+		Client:        c,
+		// TODO: temporary URL
+		URL: "https://storage.googleapis.com/scanner-v4-test/vulnerability-bundles/4.3.x-173-g6bbb2e07dc/output.json.zst",
```
This is temporary for testing purposes

RTann commented on go.mod:
```diff
@@ -84,8 +85,9 @@ require (
 	github.com/prometheus/client_model v0.5.0
 	github.com/prometheus/common v0.45.0
-	github.com/quay/claircore v1.5.19
+	github.com/quay/claircore v1.5.21-0.20231219014855-9210797e665e
```
switched to `main` branch for testing purposes

jvdm replied:
> switched to main branch for testing purposes

What did you need from main? If there is a chance it is going to be required for our release, maybe leave a comment.

RTann replied:
I definitely expect us to use an official ClairCore release for our release. v1.5.19 is several months old at this point, and we can't use v1.5.20 because it doesn't build on macOS. The main branch fixes that build issue and also has many updates. One of those is the newer way of fetching layers which had that memory issue (so it was good I tested with the main branch 😄)

RTann commented on scanner/Makefile:
```diff
@@ -165,6 +161,19 @@ copy-scripts: $(addprefix ../image/rhel/static-bin/,$(SCRIPTS))
+#################
+
+integration-d       := matcher/metadata/postgres
+integration-go-tag  := scanner_integration
```
used `scanner_integration` instead of `integration` to make sure it doesn't collide with `integration` as used outside the `scanner` directory

jvdm replied:
Nitpick: Maybe a Makefile comment instead of a PR comment.

RTann commented on scanner/matcher/matcher.go:
```diff
@@ -40,43 +42,68 @@ func NewMatcher(ctx context.Context, cfg config.MatcherConfig) (Matcher, error)
+		return nil, fmt.Errorf("creating vuln updater: %w", err)
+	}
+
+	go func() {
```
This allows the matcher to be used even without vulnerabilities. We need to decide how to determine when the matcher is `ready`

RTann commented on scanner/matcher/updater/updater.go:
```diff
@@ -0,0 +1,326 @@
+	skipGC          bool
+	updateRetention int
+
+	importVulns func(ctx context.Context, reader io.Reader) error
```
This is for testing purposes. See the test file

jvdm commented on scanner/matcher/updater/updater.go:
```diff
@@ -0,0 +1,326 @@
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			zlog.Error(ctx).Err(err).Msg("error closing temp update file")
```
Nit: The "error" in "error closing..." is redundant. We're emitting an error-level log entry.

jvdm commented on scanner/services/matcher_test.go:
```diff
@@ -125,3 +125,21 @@ func (s *matcherServiceTestSuite) Test_matcherService_GetVulnerabilities_empty_c
+	protoNow, err := types.TimestampProto(now)
+	s.Require().NoError(err)
+
+	s.matcherMock.
+		EXPECT().
+		GetLastVulnerabilityUpdate(gomock.Any()).
+		Return(now, nil)
```
A `Return(nil, err)` scenario is worth adding.

jvdm commented on scanner/services/matcher.go:
```diff
@@ -98,8 +96,19 @@ func (s *matcherService) parseIndexReport(contents *v4.Contents) (*claircore.Ind
+		return nil, fmt.Errorf("getting last vulnerability update time: %w", err)
+	}
+
+	timestamp, err := types.TimestampProto(lastVulnUpdate)
```
Sanity: It might not be relevant if the first update is part of the readiness check, but anyway, while the first update runs, what does this conversion yield?

jvdm commented on scanner/matcher/matcher.go:
```diff
@@ -40,43 +42,68 @@ func NewMatcher(ctx context.Context, cfg config.MatcherConfig) (Matcher, error)
+
 // Close closes the matcher.
 func (m *matcherImpl) Close(ctx context.Context) error {
-	ctx = zlog.ContextWithValues(ctx, "component", "scanner/backend/matcher")
+	ctx = zlog.ContextWithValues(ctx, "component", "scanner/backend/matcher.Close")
 	return m.libVuln.Close(ctx)
 }
```
We also need to gracefully shut down the updater.

jvdm commented on scanner/matcher/matcher.go:
```diff
@@ -40,43 +42,68 @@ func NewMatcher(ctx context.Context, cfg config.MatcherConfig) (Matcher, error)
 	// TODO: Update HTTP client.
 	c := http.DefaultClient
+
 	libVuln, err := libvuln.New(ctx, &libvuln.Options{
```
Maybe we want to wait for at least one successful updater run before we say "libvuln is ready". That would help with the initialization part. Fine to leave this outside the PR.

jvdm reviewed (commented):
I ignored the PSQL aspects of this patch for now, sending a preliminary review to send and get some feedback earlier. But generally, LGTM. Great work.

RTann commented on scanner/matcher/updater/updater.go:
I saw these as different enough for this to be ok. This is the specific matcher-related updater, which downloads the vuln .json.zst file and inserts it all into the DB. This differs from `scanner/updater` which downloads from the internet and creates the .json.zst file.

> they do have common data structures and some coupled logic

which parts, specifically?

daynewlee replied:
> Nitpick: A little bit confusing to have `scanner/updater` and `scanner/matcher/updater`; they do have common data structures and some coupled logic, might be worth having them in the same package.

I think it would be a bit confusing if we have `scanner/updater` with `scanner/matcher/updater` in the same package since one is used in CI and the other is designed specifically for matcher.

jvdm replied:
> I saw these as different enough for this to be ok. This is the specific matcher-related updater, which downloads the vuln .json.zst file and inserts it all into the DB. This differs from scanner/updater which downloads from the internet and creates the .json.zst file.

What I mean is, the `scanner/matcher/updater` + `scanner/updater` is akin to a ClairCore updater: Pull vulns from security sources and make them available to the matcher DB. It's just that our updater store the data in a temporary remote location. But the data and types are coupled -- in the sense the format of the archives, names, etc, need to be in sync.

This is nit, I'm fine going with what you're proposing.

RTann replied:
I'm going to keep it here for now, but if it's confusing later, we can move it

daynewlee commented on scanner/matcher/metadata/migrations/01-init.sql:
```diff
@@ -0,0 +1,6 @@
+--- a table to hold the timestamp for the latest vulnerability update.
```
Have we considered to store this information in memory but not in DB? Are there any concerns regarding memory consumption or issues that might arise during pod restarts?

RTann replied:
This is stored in the DB for two main reasons:

1. To share it among all replicas. It's possible different replicas will perform the update process, so each one will need to know the previous timestamp. Storing it in the DB allows us to share it among all replicas.
2. Persistence. This can help us startup faster, as we can read this from the DB to know that we already have vulns, so no need to run the update process immediately nor block scans from succeeding

dcaravel commented on scanner/matcher/updater/updater.go:
```diff
@@ -0,0 +1,366 @@
+			zlog.Info(ctx).Msg("completed update")
+
+			timer.Reset(u.updateInterval + jitter())
+		case <-ctx.Done():
+			return ctx.Err()
```
This doesn't appear reachable - perhaps I'm missing it but not seeing anything close/cancel the context. That by design or is there a timeout / cancel / close expected somewhere?

dcaravel replied:
Ahh there it is, excellent.

dcaravel reviewed (commented):
nit: Is there a JIRA ticket for this feature? If so add to title of PR.

RTann commented on scanner/matcher/metadata/migrations/01-init.sql:
```diff
@@ -0,0 +1,6 @@
+    id SERIAL PRIMARY KEY,
+    key VARCHAR(128) NOT NULL UNIQUE,
+    timestamp TEXT
+);
```
I don't think ClairCore assumes control over the whole DB, just the tables it creates. To be safe, I asked the Clair team, and I will followup once I hear back

RTann replied:
I'll create a ticket to address this. This may be pretty involved

RTann replied:
https://github.com/stackrox/stackrox/pull/9354

RTann commented on scanner/matcher/metadata/migrations/01-init.sql:
```diff
@@ -0,0 +1,6 @@
+    id SERIAL PRIMARY KEY,
+    key VARCHAR(128) NOT NULL UNIQUE,
+    timestamp TEXT
+);
```
> Have we considered a generic table for key/value pairs? Could make things easier in the future when need to store random values for whatever reason.

thought about it, as we have this in Scanner v2. I decided, instead, to keep separate things separate. Yes, that means potentially more tables, but it also makes it easier to have more control over each specific thing we care about

> What is the role of the id serial/sequence? If key is unique we wouldn't have two rows with the same key (if I understand correctly), wondering if id is necessary or could be removed.

remnant from the original idea of using a generic key/value table. I guess it's not needed anymore

RTann commented on scanner/matcher/matcher.go:
```diff
@@ -41,64 +45,71 @@ func NewMatcher(ctx context.Context, cfg config.MatcherConfig) (Matcher, error)
-		},
-		UpdaterSets: updaterSets,
-		Updaters:    oot,
+		// TODO: Replace with our own enrichers.
```
ticket exists, but I can't recall which one it is. I will update once I find it

dcaravel reviewed (approved):
LGTM, is there an high-level ticket for this changes? if so add to PR title.

RTann commented:
@jvdm @daynewlee @dcaravel Apologies for this, but I made some significant updates since you may have reviewed, so I'm going to ask for a re-review

The updates include:

* moving the database changes to `scanner/datastore/postgres`. This gives us a unified area to put all future DB things. Influenced by ClairCore
* Updates to GC operations
  * I found when we run the GC, the number of remaining items still increases. ClairCore throttles a single GC run to delete 50 update operations, and our update process adds more than that (I think maybe around 70 or so?), so our database size will continue to increase over time. So, I added a periodic "full GC" process to run about once per day
  * Instead of running the typical, single GC process asynchronously, run it synchronously. This prevents the GC process from potentially slowing down any subsequent updates, should they run at the same time. ClairCore does this, too

daynewlee commented:
> nit: Is there a JIRA ticket for this feature? If so add to title of PR.

I think it's part of 19005 and part of scale testing ticket (don't know the number)

RTann commented on scanner/services/matcher.go:
```diff
@@ -112,8 +110,19 @@ func (s *matcherService) parseIndexReport(contents *v4.Contents) (*claircore.Ind
+		return nil, fmt.Errorf("internal error: %w", err)
+	}
+	return &v4.Metadata{
+		LastVulnerabilityUpdate: timestamp,
```
TODO: if https://github.com/stackrox/stackrox/pull/9355 is merged first, be sure to set the scanner version

dcaravel reviewed (approved):
LGTM, few nits/comments. FYI, few CI style check errors.

RTann commented on scanner/matcher/matcher.go:
```diff
@@ -2,93 +2,153 @@ package matcher
+	}
+
+	// matcherNames specifies the ClairCore matchers to use.
+	// TODO: add NodeJS once implemented.
```
done https://github.com/stackrox/stackrox/pull/9042/commits/037a6a00c8ac52fc512045a43e6f89bef82d9d0c

RTann commented on scanner/matcher/matcher.go:
```diff
@@ -2,93 +2,153 @@ package matcher
 	if err != nil {
 		return nil, fmt.Errorf("initializing postgres matcher store: %w", err)
 	}
+	metadataStore, err := postgres.InitPostgresMatcherMetadataStore(ctx, pool, true)
```
Once https://github.com/stackrox/stackrox/pull/9354 is merged, I plan to use a separate pool

RTann commented on scanner/matcher/updater/updater.go:
```diff
@@ -0,0 +1,356 @@
+
+	if !u.skipGC {
+		go u.runGCFullPeriodic(ctx)
+	}
```
I think inside `Start` makes more sense. I'll put it there, thanks

RTann replied:
moved it https://github.com/stackrox/stackrox/pull/9042/commits/09fd185a96871db8a02faae8ca37ea9ca1ebde00

RTann commented:
@dcaravel just a heads up I pushed a few more commits after your review, but they were very minimal. Notable changes:

* added comments
* removed the explicit updater set for libvuln because it's not used...
* added a for loop inside the periodic GC so it'd actually be periodic

RTann commented on scanner/matcher/updater/gc.go:
```diff
@@ -0,0 +1,98 @@
+		case <-t.C:
+			zlog.Info(ctx).Msg("full GC started")
+			u.runGCFull(ctx)
+			zlog.Info(ctx).Msg("full GC completed")
```
Put duration in the log?

RTann commented on scanner/matcher/updater/gc.go:
```diff
@@ -0,0 +1,98 @@
+	zlog.Info(ctx).
+		Int64("remaining_ops", i).
+		Int("retention", u.updateRetention).
+		Msg("GC completed")
```
Put duration in the log?

RTann commented on scanner/matcher/updater/updater.go:
```diff
@@ -0,0 +1,364 @@
+
+// update runs the full vulnerability update process.
+func (u *Updater) update(ctx context.Context) error {
+	if err := u.runUpdate(ctx); err != nil {
```
consider returning a bool to indicate if the update actually happened. perhaps don't bother running GC if not update occurred
